### PR TITLE
Add `SklearnTrainigSession` for sklearn autologging

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

Add `SklearnTrainigSession` for sklearn autologging. 

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/sklearn/test_sklearn_training_session.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import pytest
```
Comment (harupy) [inline tests/sklearn/test_sklearn_training_session.py:1]: Created a file that focuses on `_SklearnTrainingSession` to simplify test function names

Comment (harupy) [reply tests/sklearn/test_sklearn_training_session.py:1]: I'll add more tests

Comment (dbczumar) [reply tests/sklearn/test_sklearn_training_session.py:1]: Looks great! Thanks so much. On the subject of additional testing, it'd be awesome to include a case verifying that, if we create a session with `allow_children=False` and a nested session with `allow_children=True`, additional nested sessions should still not log (i.e., the top-level session's `allow_children=True` should override the second level's `allow_children=False`).

File: mlflow/sklearn.py [hunk 2]
```diff
@@ -424,3 +424,29 @@ def load_model(model_uri):
     return _load_model_from_local_file(
         path=sklearn_model_artifacts_path, serialization_format=serialization_format
     )
+
+
+class _SklearnTrainingSession(object):
```
Comment (dbczumar) [inline mlflow/sklearn.py:429]: Can we document the class constructor with information about the `clazz` and `allow_children` parameters & also document the `should_log()` method with information about the behavior?

Comment (dbczumar) [inline mlflow/sklearn.py:429]: Can we include some information about the lack of thread safety guarantees for this class and why that's okay for now? I.e., 1. we don't currently have any use cases for `allow_children=True` and 2. the list `append` & `pop` operations are thread-safe, so we will always clear the session stack once all `SklearnTrainingSessions` exit.

In the future, we'll need to figure out how to handle multithreaded backends.

File: mlflow/sklearn.py [hunk 3]
```diff
@@ -424,3 +424,46 @@ def load_model(model_uri):
...
+    _session_stack = []
+
+    def __init__(self, clazz, allow_children=True):
+        """
+        A session manager for nested autologging runs.
+
+        :param clazz: A class object that this session originates from.
+        :param allow_children: If True, allows autologging in child sessions.
+                               If False, disallows autologging in all descendant sessions.
```
Comment (harupy) [inline mlflow/sklearn.py:442]: @dbczumar What do you think about providing some examples here like below?

```python

>>> with _SklearnTrainingSession(Parent, False) as p:
...     with _SklearnTrainingSession(Child, True) as c:
...         with _SklearnTrainingSession(Grandchild, True) as g:
...             print(p.should_log())  # -> True
...             print(c.should_log())  # -> False
...             print(g.should_log())  # -> False

>>> with _SklearnTrainingSession(Parent, True) as p:
...    with _SklearnTrainingSession(Child, False) as c:
...        with _SklearnTrainingSession(Grandchild, True) as g:
...            print(p.should_log())  # -> True
...            print(c.should_log())  # -> True
...            print(g.should_log())  # -> False
```

Comment (dbczumar) [reply mlflow/sklearn.py:442]: Sure! An example like this seems valuable to include

**Review by dbczumar — APPROVED** [review]
LGTM! (pending optional inclusion of the example described in https://github.com/mlflow/mlflow/pull/3263/files#r470774807). Thanks @harupy!

File: mlflow/sklearn.py [hunk 4]
```diff
@@ -424,3 +424,73 @@ def load_model(model_uri):
...
+
+        >>> with _SklearnTrainingSession(Parent, False) as p:
+        ...     with _SklearnTrainingSession(Child, True) as c:
+        ...         with _SklearnTrainingSession(Grandchild, True) as g:
+        ...             print(p.should_log())
+        ...             print(c.should_log())
+        ...             print(g.should_log())
+        True
+        False
+        False
```
Comment (harupy) [inline mlflow/sklearn.py:458]: Used doctest syntax here so we can test this example by running:

```
pytest --doctest-modules mlflow/sklearn.py
```

**Pull request merged**